### DISCUSSION TITLE
Make patches discard build cache (or allow config spec)

### DISCUSSION THREAD

**Discussion by haggholm** [vercel/community#1844] [Idea]
Node project. One recurring problem I have found is that whenever we modify NPM patches, Vercel deployments fail. I presume this is because the build cache stores the _patched_ version of a library in `node_modules`, so when a solution like `patch-package` attempts to apply a clean patch, it ends up trying to apply to an already-patched library and fails.

Obviously, I can fix this by going to the Vercel UI and manually triggering a redeploy without reusing build cache; but this introduces manual steps, and gets somewhat complicated when it also involves CI pipelines, and when merges and rebasing means that the same patch change may end up requiring a number of manual interventions on different branches (including for people unaware of the change).

It would be very nice if this didn't happen—and it is predictable based on the state of the repository compared to the previous run. This could be done by Vercel being aware of common directories (like `./.yarn/patches`, `./patches`, etc.) that, when changed, are likely to invalidate build state. Alternatively, there could be a configuration option to allow users to specify the same thing.

**Comment by amyegan** [comment]
Thanks for sharing this idea! Can you share more about your typical workflow? I want to pass the right information along to the team.

I also want to figure out if we already have a solution that might make this process easier for you. For example, you can use a [`VERCEL_FORCE_NO_BUILD_CACHE`](https://vercel.com/docs/concepts/deployments/troubleshoot-a-build#managing-build-cache) environment variable to bypass the build cache. Or, if you use the CLI to deploy, the [`--force`](https://vercel.com/docs/cli/global-options#force) option can be used to skip the build cache.

**Comment by haggholm** [comment]
Thank you for the quick response!

Our team works on a decently sized monorepo with various components for our software stack, and uses NPM workspaces, including a few patches. Vercel is watching our git repo and creates builds whenever we push to a branch, as part of our CI process. In other words, we don't use the CLI tool or explicitly interact with Vercel; we just push to our git repo and, 99% of the time, it Just Works™.

What most recently happened was that I (a backend developer) modified a patch for a buggy NPM library using a tool called [patch-package](https://www.npmjs.com/package/patch-package), which stores patches in the `./patches` directory of our monorepo (as we use NPM—Yarn has something similar built-in via [yarn patch](https://yarnpkg.com/cli/patch)); once this was merged, various other people got the patch by merging or rebasing on top of my commit. Since the integration works via Vercel watching our repository, there's no CLI involved and no opportunity I can see to modify environment variables. 

But unfortunately, every branch that gets the patch modification fails, because `patch-package` cannot cleanly apply the patch; since it runs in the `postinstall` phase, this means `npm ci` fails. When this happens, you have to manually interact with Vercel, which is not usually part of our workflow, to re-run the deployment without build cache. And since many branches will already have build caches, this ends up impacting numerous people every time such a change to patches is made. Worse, it's not terribly obvious to everyone why they're failing or what they need to do. Obviously, the fix itself is quite trivial (go to Vercel, hit Redeploy, keep the build cache off—which is the default already when redeploying a failed build); the cost is in human overhead and confusion; on our side it ends up being poor dev UX.

I _think_ this might be an issue only when modifying, rather than initially adding, patches (if you just add one, it ought to cleanly apply), but the loss of build performance from being overly aggressive in pruning build cache from changes to `./patches` is insignificant compared to the nuisance of having builds fail and needing people to figure out what's going on and how to resolve it. So if I could just update our `vercel.json` to add an entry like (pulling a bad name out of thin air) `{"invalidateCacheDirectories:["patches"]}`, I think it would avoid a lot of headaches and confusion.

**Reply by amyegan** [reply]
Beautiful explanation. Thanks for writing all of this! I shared it with the engineering team and I'll let you know when I have more information

**Reply by smaeda-ks** [reply]
@haggholm could adding a mini script and running it right before your install command be a workaround?
e.g., customize your install command; `./patch-check.sh && yarn install`

```sh
#!/bin/sh

path="./patches"
git diff HEAD^ HEAD --quiet ${path}
status=$?

if [ $status -eq 0 ]; then
    echo "No patches to apply"
else
    echo "New patches available, removing build cahces..."
    rm -rf node_modules
    rm -rf .next/cache
    echo "Removed build caches"
fi

exit 0
```
